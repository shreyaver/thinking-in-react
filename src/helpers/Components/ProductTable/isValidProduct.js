const isValidProduct = (inStockOnly, filterText, productData) => {
  if ((inStockOnly === false) || (inStockOnly === true && productData.stocked === true)) {
    if (productData.name.indexOf(filterText) !== -1) {
      return true;
    }
  }
  return false;
};
export default isValidProduct;
