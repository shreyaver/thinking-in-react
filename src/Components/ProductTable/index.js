import React from 'react';
import PropTypes from 'prop-types';
import ProductCategoryRow from '../ProductCategoryRow';
import ProductRow from '../ProductRow';
import isValidProduct from '../../helpers/Components/ProductTable/isValidProduct';

const ProductTable = (props) => {
  let currentCategory = null;
  const rows = props.productArray.reduce((result, productData) => {
    if (productData.category !== currentCategory) {
      result.push(<ProductCategoryRow key={productData.category} category={productData.category} />);
      currentCategory = productData.category;
    }
    if (isValidProduct(props.inStockOnly, props.filterText, productData)) {
      result.push(<ProductRow key={productData.name} product={productData} />);
    }
    return result;
  }, []);
  const tableStyle = { border: '1px solid black' };
  return (
    <table style={tableStyle}>
      <thead>
        <tr>
          <th>Name</th>
          <th>Price</th>
        </tr>
      </thead>
      <tbody>{rows}</tbody>
    </table>
  );
};
ProductTable.propTypes = {
  productArray: PropTypes.array.isRequired,
  filterText: PropTypes.string,
  inStockOnly: PropTypes.bool,
};
ProductTable.defaultProps = {
  filterText: '',
  inStockOnly: false,
};
export default ProductTable;
