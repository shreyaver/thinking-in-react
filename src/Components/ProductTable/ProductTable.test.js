import React from 'react';
import renderer from 'react-test-renderer';
import ProductTable from '.';

describe('ProductTable', () => {
  const productArray = [
    {
      category: 'Sporting Goods', price: '$49.99', stocked: true, name: 'Football',
    },
    {
      category: 'Sporting Goods', price: '$9.99', stocked: true, name: 'Baseball',
    },
    {
      category: 'Sporting Goods', price: '$29.99', stocked: false, name: 'Basketball',
    },
    {
      category: 'Electronics', price: '$99.99', stocked: true, name: 'iPod Touch',
    },
    {
      category: 'Electronics', price: '$399.99', stocked: false, name: 'iPhone 5',
    },
    {
      category: 'Electronics', price: '$199.99', stocked: true, name: 'Nexus 7',
    },
  ];
  it('renders without crashing', () => {
    const productTable = renderer.create(<ProductTable productArray={productArray} />).toJSON();
    expect(productTable).toMatchSnapshot();
  });
});
// xit('renders without crashing for only stocked products and no filter text', () => {
//   const productTable = renderer.create(<ProductTable productArray = {productArray} filterText= '' inStockOnly= {true} />).toJSON();
//   expect(productTable).toMatchSnapshot();
// });
// xit('renders without crashing for stocked and unstocked products and passed filter text', () => {
//   const productTable = renderer.create(<ProductTable productArray = {productArray} filterText= 'Phone' inStockOnly= {false} />).toJSON();
//   expect(productTable).toMatchSnapshot();
// });
// xit('renders without crashing for only stocked products and passed filter text', () => {
//   const productTable = renderer.create(<ProductTable productArray = {productArray} filterText= 'ball' inStockOnly= {true} />).toJSON();
//   expect(productTable).toMatchSnapshot();
// });
