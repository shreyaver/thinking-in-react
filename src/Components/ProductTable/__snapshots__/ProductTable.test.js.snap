// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTable renders without crashing 1`] = `
<table
  style={
    Object {
      "border": "1px solid black",
    }
  }
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Price
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        colSpan="2"
      >
        Sporting Goods
      </th>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "black",
          }
        }
      >
        Football
      </td>
      <td>
        $49.99
      </td>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "black",
          }
        }
      >
        Baseball
      </td>
      <td>
        $9.99
      </td>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "red",
          }
        }
      >
        Basketball
      </td>
      <td>
        $29.99
      </td>
    </tr>
    <tr>
      <th
        colSpan="2"
      >
        Electronics
      </th>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "black",
          }
        }
      >
        iPod Touch
      </td>
      <td>
        $99.99
      </td>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "red",
          }
        }
      >
        iPhone 5
      </td>
      <td>
        $399.99
      </td>
    </tr>
    <tr>
      <td
        className="ProductName"
        style={
          Object {
            "color": "black",
          }
        }
      >
        Nexus 7
      </td>
      <td>
        $199.99
      </td>
    </tr>
  </tbody>
</table>
`;
